### ISSUE TITLE
Task2 Bayes_plus

### ISSUE THREAD

**Issue by Zdhj** [Zdhj/Blog#3]
# Overview
## Content
All the [tasks](https://github.com/datawhalechina/team-learning/tree/master/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E7%AE%97%E6%B3%95%E5%9F%BA%E7%A1%80) are listed here.
[Task2 Bayes_plus](https://github.com/datawhalechina/team-learning/blob/master/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E7%AE%97%E6%B3%95%E5%9F%BA%E7%A1%80/Task2%20bayes_plus.ipynb)

## Tools and knowledge
 _sklearn_ is a is a Python module for **machine learning**. So how to use it? 
In [Chinese](https://sklearn.apachecn.org/).
With [introduction](https://www.jianshu.com/p/516f009c0875).
With [a easy example](https://blog.csdn.net/lawme/article/details/78726643).

# Error found and experience in [Task2 Bayes_plus](https://github.com/datawhalechina/team-learning/blob/master/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E7%AE%97%E6%B3%95%E5%9F%BA%E7%A1%80/Task2%20bayes_plus.ipynb)
## Type of model in Probability and Statistics Theory
_Generative model_ refers to a model that can randomly generate observation data, which mainly includes **Gaussian mixture model** and other mixture models, **hidden Markov model**, **random context-free grammar**, **naive Bayes classifier**, **AODE classifier**, **latent Dirichlet allocation model**, **restricted Boltzmann machine**.

_Discriminant model_ refers to a model that finds the relationship between unknown data and known data, which mainly includes **logistic regression**, **linear regression**, **support vector machine**, **lifting method**, **conditional random field**, **artificial neural network**, **random forest**, **perceptron**.

## Probability
_Conditional Probability_ refers to the probability that event _A_ occurs under the condition that event _B_ occurs.
_Priori probability_ refers to the probability that event _B_ occurs under previous experience and analysis.
_Posterior probability_ refers to the probability that event _B_ occurs under the condition that event _A_ occurs

## Bayesian decision theory
_Bayesian criterion_
_Bayesian optimal classifier_
_Bayesian risk_

## Bayesian formula
![image](https://user-images.githubusercontent.com/25249526/80081697-dd68d100-8585-11ea-9834-b999a876f1fb.png)
where, 
* _P(c)_ is the priori probability,
* _P(x|c)_ is the **class conditional probability** of the sample _x_ for the class label _c_, 
* _P(x)_ is the "evidence" factor used for normalization.

It is worth noting that _P(x|c)_ is difficult to be estimated directly from limited training samples.
* Combination explosion problem may happen;
* Sample sparseness problem may happen;
* Problem may be more serious with more attributes.

### Maximum likelihood estimation
Maximum likelihood estimation is a common strategy for estimating conditional probability,which has a **certain form** of probability distribution, and then estimate the **parameters** of the probability distribution based on training samples.
![image](https://user-images.githubusercontent.com/25249526/80085361-b1038380-858a-11ea-95db-2615fb7ae45f.png)

It is worth noting that the accuracy of the estimation results depends heavily on whether the assumed form of probability distribution conforms to the underlying real data distribution.

Steps
1. Write out the likelihood function;
2. Take the logarithm of the likelihood function;
3. Derivative the likelihood function so that the partial derivative is 0 to obtain the likelihood equation equations;
4. Solve the likelihood equations and get all the parameters.

## Naive Bayes classifier
_Attribute conditional independence assumption_ refers to assuming that all attributes of known categories are independent of each other.
![image](https://user-images.githubusercontent.com/25249526/80095793-1bbcbb00-859b-11ea-8a87-58b6b825c8dc.png)

## Underflow problem
Underflow problem refers to that the calculation result of the floating point number of the computer is less than the minimum number that can be represented under computer's ability.
When the value is less than a certain number, it cannot be accurately saved, which will cause the loss of precision.

## Zero probability problem
During the actual model training process, the _priori probability_ and the _anti-conditional probability_ are calculated based on the training sample. 

But the number of training samples is not unlimited, it may happen that some situations exist in practice, not in the training samples, resulting in a probability value of 0, affecting the calculation of _Posterior probability_. 

We call it unsmooth model and that's why we apply _Bayesian estimation_ to train samples. And that's what we do to Naive Bayes classifier with the method of _Laplace smoothing_.

_Laplace smoothing_ 
* Add **1** to the numerator of the _prior probability_;
* Add the number of categories to the the denominator of _prior probability_;
* Add **1** to the numerator of the _Conditional Probability_;
* Add the number of possible values of features to the the denominator of the _Conditional Probability_.
![image](https://user-images.githubusercontent.com/25249526/80094717-558cc200-8599-11ea-83da-a9736fb59813.png)
where, 
* _N_ is the category label of data set,
* _Ni_ is the number of value categories of the _i_ attribute, 
* _|D|_ is the size of samples,
* _Dc_ is the number of type _c_ recorded 
* _|Dxi|c|_ is the number of the _i_ attribute with the value of _xi_ in type _c_.
